### PR TITLE
User micro

### PR DESCRIPTION
**Pull request by tranduytung99**

### CODE REVIEW

File: app/controllers/microposts_controller.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
+class MicropostsController < ApplicationController
+  before_action :logger_in_used, only: [:create, :destroy]
+  before_action :correct_user, only: :destroy
+  
+  def create
+  	@micropost = current_user.microposts.build(micropost_params)
```
Comment (hungnh103) [inline app/controllers/microposts_controller.rb]: * indent
* remove `()`

File: app/controllers/microposts_controller.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    if @micropost.save
+      flash[:success] = "Micropost created!"
```
Comment (hungnh103) [inline app/controllers/microposts_controller.rb]: I18n

File: app/controllers/microposts_controller.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+      redirect_to root_url
+    else
+      @feed_items = []
+      render "static_pages/home"
+    end
+  end
+
+  def destroy
+    @micropost.destroy
+    flash[:success] = "Deleted"
```
Comment (hungnh103) [inline app/controllers/microposts_controller.rb]: I18n

File: app/controllers/microposts_controller.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    redirect_to request.referrer || root_url
+  end
+
+  private
+
+    def micropost_params
+      params.require(:micropost).permit(:content, :picture)
```
Comment (hungnh103) [inline app/controllers/microposts_controller.rb]: remove `()`

File: app/controllers/static_pages_controller.rb [hunk 5]
```diff
@@ -1,5 +1,9 @@
 class StaticPagesController < ApplicationController
   def home
+    if logged_in?
+      @micropost  = current_user.microposts.build
```
Comment (hungnh103) [inline app/controllers/static_pages_controller.rb]: remove whitespace

File: app/models/micropost.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
+class Micropost < ApplicationRecord
+  belongs_to :user
+  mount_uploader :picture, PictureUploader
+  validates :user_id, presence: true
+  validates :content, presence: true, length:{ maximum: 140 }
```
Comment (hungnh103) [inline app/models/micropost.rb]: * convention
* use gem `config`

File: app/models/micropost.rb [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+  validate :picture_size
+
+  private
+
+  def picture_size
+    if picture.size > 10.megabytes
```
Comment (hungnh103) [inline app/models/micropost.rb]: use gem `config`

File: app/models/micropost.rb [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+      errors.add(:picture , "max 10")
```
Comment (hungnh103) [inline app/models/micropost.rb]: * remove `()`?
* I18n

File: app/models/user.rb [hunk 9]
```diff
@@ -34,6 +35,10 @@ def authenticated? attribute, token
     BCrypt::Password.new(digest).is_password?(token)
   end
 
+  def feed
+    Micropost.where("user_id = ?", id)
```
Comment (hungnh103) [inline app/models/user.rb]: remove `()`

File: app/uploaders/picture_uploader.rb [hunk 10]
```diff
@@ -0,0 +1,53 @@
+# encoding: utf-8
+
+class PictureUploader < CarrierWave::Uploader::Base
+  include CarrierWave::MiniMagick
+  process resize_to_limit: [400, 400]
```
Comment (hungnh103) [inline app/uploaders/picture_uploader.rb]: use gem `config`

File: app/views/microposts/_micropost.html.erb [hunk 11]
```diff
@@ -0,0 +1,14 @@
+<li id="micropost-<%= micropost.id %>">
+  <%= link_to gravatar_for(micropost.user, size: 50), micropost.user %>
```
Comment (hungnh103) [inline app/views/microposts/_micropost.html.erb]: use gem `config`

File: app/views/microposts/_micropost.html.erb [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+  <span class="user"><%= link_to micropost.user.name, micropost.user %></span>
+  <span class="content"><%= micropost.content %></span>
+  <% if micropost.picture?%>
+  <%= image_tag micropost.picture.url, :class => "img_tag_micro" %>
```
Comment (hungnh103) [inline app/views/microposts/_micropost.html.erb]: `class: "img_tag_micro"`

File: app/views/microposts/_micropost.html.erb [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+  <% end %>
+  <span class="timestamp">
+    Posted <%= time_ago_in_words(micropost.created_at) %> ago.
```
Comment (hungnh103) [inline app/views/microposts/_micropost.html.erb]: * I18n
* remove `()` ?

File: app/views/microposts/_micropost.html.erb [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+    <% if current_user?(micropost.user) %>
```
Comment (hungnh103) [inline app/views/microposts/_micropost.html.erb]: remove `()`

File: app/views/microposts/_micropost.html.erb [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+      <%= link_to "Delete", micropost, method: :delete, data:{confirm: "You sure?"} %>
```
Comment (hungnh103) [inline app/views/microposts/_micropost.html.erb]: * I18n
* missing whitespace

File: app/views/shared/_micropost_form.html.erb [hunk 16]
```diff
@@ -0,0 +1,10 @@
+<%= form_for @micropost do |f| %>
+  <%= render 'shared/error_messages', object: f.object %>
```
Comment (hungnh103) [inline app/views/shared/_micropost_form.html.erb]: use double quote `" "` in Rails

File: app/views/shared/_micropost_form.html.erb [hunk 17]
```diff
@@ -0,0 +1,10 @@
...
+  <div class="field">
+    <%= f.text_area :content, placeholder: "Compose new micropost..." %>
```
Comment (hungnh103) [inline app/views/shared/_micropost_form.html.erb]: I18n

File: app/views/shared/_micropost_form.html.erb [hunk 18]
```diff
@@ -0,0 +1,10 @@
...
+  </div>
+  <%= f.submit "Post", class: "btn btn-primary" %>
```
Comment (hungnh103) [inline app/views/shared/_micropost_form.html.erb]: I18n

File: app/views/shared/_micropost_form.html.erb [hunk 19]
```diff
@@ -0,0 +1,10 @@
...
+    <span class="picture">
+    <%= f.file_field :picture %>
+  </span>
+<% end %>
```
Comment (hungnh103) [inline app/views/shared/_micropost_form.html.erb]: missing end file

File: app/views/shared/_user_info.html.erb [hunk 20]
```diff
@@ -0,0 +1,4 @@
+<%= link_to gravatar_for(current_user, size: 50), current_user %>
```
Comment (hungnh103) [inline app/views/shared/_user_info.html.erb]: use gem `config`

File: app/views/shared/_user_info.html.erb [hunk 21]
```diff
@@ -0,0 +1,4 @@
...
+<h1><%= current_user.name %></h1>
+<span><%= link_to "view my profile", current_user %></span>
+<span><%= pluralize(current_user.microposts.count, "micropost") %></span>
```
Comment (hungnh103) [inline app/views/shared/_user_info.html.erb]: * remove `()`
* missing end file

File: app/views/static_pages/home.html.erb [hunk 22]
```diff
@@ -1,4 +1,21 @@
 <% provide :title, t(".title") %>
+<% if logged_in? %>
+  <div class="row">
+    <aside class="col-md-4">
+      <section class="user_info">
+        <%= render 'shared/user_info' %>
```
Comment (hungnh103) [inline app/views/static_pages/home.html.erb]: use double quote `" "` in Rails

File: app/views/static_pages/home.html.erb [hunk 23]
```diff
@@ -1,4 +1,21 @@
...
+      </section>
+      <section class="micropost_form">
+        <%= render 'shared/micropost_form' %>
```
Comment (hungnh103) [inline app/views/static_pages/home.html.erb]: use double quote `" "` in Rails

File: app/views/static_pages/home.html.erb [hunk 24]
```diff
@@ -1,4 +1,21 @@
...
+      </section>
+    </aside>
+    <div class="col-md-8">
+      <h3>Micropost Feed</h3>
```
Comment (hungnh103) [inline app/views/static_pages/home.html.erb]: I18n

File: app/views/static_pages/home.html.erb [hunk 25]
```diff
@@ -1,4 +1,21 @@
...
+      <%= render 'shared/feed' %>
```
Comment (hungnh103) [inline app/views/static_pages/home.html.erb]: use double quote `" "` in Rails

File: app/views/static_pages/home.html.erb [hunk 26]
```diff
@@ -12,3 +29,4 @@
 </div>
 
 <%= link_to image_tag("rails.png", alt: t(".logo") ),"http://rubyonrails.org/" %>
+<% end %>
```
Comment (hungnh103) [inline app/views/static_pages/home.html.erb]: missing end file

File: app/views/static_pages/home.html.erb [hunk 27]
```diff
@@ -12,3 +29,4 @@
 </div>
 
 <%= link_to image_tag("rails.png", alt: t(".logo") ),"http://rubyonrails.org/" %>
```
Comment (hungnh103) [inline app/views/static_pages/home.html.erb]: remove whitespace

File: app/views/users/show.html.erb [hunk 28]
```diff
@@ -9,4 +9,14 @@
       </h1>
     </section>
   </aside>
+    <div class="col-md-8">
+    <% if @user.microposts.any? %>
+      <h3>Microposts (<%= @user.microposts.count %>)</h3>
```
Comment (hungnh103) [inline app/views/users/show.html.erb]: I18n

File: app/controllers/application_controller.rb [hunk 29]
```diff
@@ -1,4 +1,15 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
   include SessionsHelper
+
+  private
+  
+  def logger_in_used
+    unless logged_in?
+      store_location
+      flash[:danger] = "Please log in"
```
Comment (huongnguyenmta) [inline app/controllers/application_controller.rb]: I18n

File: app/controllers/microposts_controller.rb (hunk 1 above)
Comment (huongnguyenmta) [inline app/controllers/microposts_controller.rb]: remove ()

**Pull request merged**